core/src/test/scala/unit/kafka/utils/TestUtils.scala

```diff
@@ -1230,7 +1230,7 @@ object TestUtils extends Logging {
     TestUtils.waitUntilTrue(
       () => {
         brokers.forall { broker =>
-          val metadataOffset = broker.asInstanceOf[BrokerServer].metadataPublisher.publishedOffset
+          val metadataOffset = broker.asInstanceOf[BrokerServer].metadataPublisher.publishedEndOffset
           metadataOffset >= controllerOffset
```

Comment:
`controllerOffset` is the last offset in the controller, inclusive. metadataOffset is exclusive. Maybe the best change is to have this line instead in line 1229:
```scala
val controllerEndOffset = controllerServer.raftManager.replicatedLog.endOffset().offset
```

Reply:
We could also maybe rename `metadataOffset` to `metadataEndOffset`.

I think we discussed this in the past but I wonder we should use types like `Offset` and `EndOffset` instead of `long` and implement comparison logic for them. Java's future primitive classes would be ideal for this: https://openjdk.org/jeps/401